File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/LibLogger.java
package me.ycdev.android.lib.common.utils;

import android.annotation.SuppressLint;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import android.util.Log;

import java.util.Locale;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class LibLogger {
    private static final String TAG = "AndroidLib";
    private static boolean sJvmLogger = false;

    @RestrictTo(RestrictTo.Scope.SUBCLASSES)
    protected LibLogger() {
        // nothing to do
    }

    public static void enableJvmLogger() {
        sJvmLogger = true;
    }

    public static void setFileLogger(FileLogger fileLogger) {
        if (!sJvmLogger) {
            AndroidLogger.setFileLogger(fileLogger);
        }
    }

    /**
     * Log enabled by default
     */
    public static void setLogEnabled(boolean enabled) {
        if (!sJvmLogger) {
            AndroidLogger.setLogEnabled(enabled);
        }
    }

    public static boolean isLogEnabled() {
        return AndroidLogger.isLogEnabled();
    }

    public static void v(@NonNull String tag, @NonNull String msg, Object... args) {
        log(Log.VERBOSE, tag, msg, null, args);
    }

    public static void d(@NonNull String tag, @NonNull String msg, Object... args) {
        log(Log.DEBUG, tag, msg, null, args);
    }

    public static void i(@NonNull String tag, @NonNull String msg, Object... args) {
        log(Log.INFO, tag, msg, null, args);
    }

    public static void w(@NonNull String tag, @NonNull String msg, Object... args) {
        log(Log.WARN, tag, msg, null, args);
    }

    public static void w(@NonNull String tag, @NonNull String msg, @NonNull Throwable e,
            Object... args) {
        log(Log.WARN, tag, msg, e, args);
    }

    public static void w(@NonNull String tag, @NonNull Throwable e, Object... args) {
        log(Log.WARN, tag, null, e, args);
    }

    public static void e(@NonNull String tag, @NonNull String msg, Object... args) {
        log(Log.ERROR, tag, msg, null, args);
    }

    public static void e(@NonNull String tag, @NonNull String msg, @NonNull Throwable e,
            Object... args) {
        log(Log.ERROR, tag, msg, e, args);
    }

    public static void log(int level, @NonNull String tag, @Nullable String msg,
            @Nullable Throwable tr, Object... args) {
        if (sJvmLogger) {
            if (msg != null && args != null && args.length > 0) {
                msg = String.format(Locale.US, msg, args);
            }
            System.out.println("[" + tag + "] " + msg);
            if (tr != null) {
                tr.printStackTrace();
            }
        } else {
            AndroidLogger.log(level, tag, msg, tr, args);
        }
    }

    private static class AndroidLogger {
        private static boolean sLogEnabled = true;
        private static FileLogger sFileLogger;

        static void setFileLogger(FileLogger fileLogger) {
            sFileLogger = fileLogger;
        }

        /**
         * Log enabled by default
         */
        static void setLogEnabled(boolean enabled) {
            sLogEnabled = enabled;
            if (!enabled && sFileLogger != null) {
                sFileLogger.close();
            }
        }

        static boolean isLogEnabled() {
            return sLogEnabled;
        }

        static void log(int level, @NonNull String tag, @Nullable String msg,
                @Nullable Throwable tr, Object... args) {
            if (showLog(level, tag)) {
                if (msg != null && args != null && args.length > 0) {
                    msg = String.format(Locale.US, msg, args);
                }
                if (tr == null) {
                    Log.println(level, tag, msg);
                } else {
                    Log.println(level, tag, msg + '\n' + Log.getStackTraceString(tr));
                }
                logToFile(tag, msg, tr);
            }
        }

        private static boolean showLog(int level, String tag) {
            return isLoggable(tag, level) || sLogEnabled;
        }

        @SuppressLint("LogNotTimber")
        private static boolean isLoggable(String tag, int level) {
            try {
                return Log.isLoggable(tag, level);
            } catch (Exception e) {
                if (sLogEnabled) {
                    throw e;
                } else {
                    Log.e(TAG, "please check the tag length?", e);
                }
            }
            return false;
        }

        private static void logToFile(String tag, String msg, Throwable tr) {
            if (sLogEnabled && sFileLogger != null) {
                sFileLogger.logToFile(tag, msg, tr);
            }
        }

    }
}

File: baseLib/src/main/java/me/ycdev/android/lib/common/utils/LibLogger.kt
package me.ycdev.android.lib.common.utils

import android.annotation.SuppressLint
import android.util.Log
import androidx.annotation.RestrictTo
import java.util.Locale

@Suppress("unused")
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
object LibLogger {
    private const val TAG = "AndroidLib"
    private var jvmLogger = false

    /**
     * Log enabled by default
     */
    var isLogEnabled: Boolean
        get() = AndroidLogger.isLogEnabled
        set(enabled) {
            if (!jvmLogger) {
                AndroidLogger.isLogEnabled = enabled
            }
        }

    fun enableJvmLogger() {
        jvmLogger = true
    }

    fun setFileLogger(fileLogger: FileLogger) {
        if (!jvmLogger) {
            AndroidLogger.setFileLogger(fileLogger)
        }
    }

    fun v(tag: String, msg: String, vararg args: Any?) {
        log(Log.VERBOSE, tag, null, msg, *args)
    }

    fun d(tag: String, msg: String, vararg args: Any?) {
        log(Log.DEBUG, tag, null, msg, *args)
    }

    fun d(tag: String, e: Throwable, msg: String, vararg args: Any?) {
        log(Log.DEBUG, tag, e, msg, *args)
    }

    fun i(tag: String, msg: String, vararg args: Any?) {
        log(Log.INFO, tag, null, msg, *args)
    }

    fun i(tag: String, e: Throwable, msg: String, vararg args: Any?) {
        log(Log.INFO, tag, e, msg, *args)
    }

    fun w(tag: String, msg: String, vararg args: Any?) {
        log(Log.WARN, tag, null, msg, *args)
    }

    fun w(tag: String, e: Throwable, msg: String, vararg args: Any?) {
        log(Log.WARN, tag, e, msg, *args)
    }

    fun w(tag: String, e: Throwable) {
        log(Log.WARN, tag, e, null)
    }

    fun e(tag: String, msg: String, vararg args: Any?) {
        log(Log.ERROR, tag, null, msg, *args)
    }

    fun e(tag: String, e: Throwable, msg: String, vararg args: Any?) {
        log(Log.ERROR, tag, e, msg, *args)
    }

    fun e(tag: String, e: Throwable) {
        log(Log.ERROR, tag, e, null)
    }

    fun log(level: Int, tag: String, tr: Throwable?, msg: String?, vararg args: Any?) {
        var msgFull = msg
        if (jvmLogger) {
            if (msgFull != null && args.isNotEmpty()) {
                msgFull = String.format(Locale.US, msgFull, *args)
            }
            println("[$tag] $msgFull")
            tr?.printStackTrace()
        } else {
            AndroidLogger.log(level, tag, tr, msgFull, *args)
        }
    }

    private object AndroidLogger {
        /**
         * Log enabled by default
         */
        internal var isLogEnabled = true
            set(enabled) {
                field = enabled
                if (!enabled && fileLogger != null) {
                    fileLogger!!.close()
                }
            }
        private var fileLogger: FileLogger? = null

        fun setFileLogger(fileLogger: FileLogger) {
            this.fileLogger = fileLogger
        }

        fun log(level: Int, tag: String, tr: Throwable?, msg: String?, vararg args: Any?) {
            var msgFull = msg
            if (showLog(level, tag)) {
                if (msgFull != null && args.isNotEmpty()) {
                    msgFull = String.format(Locale.US, msgFull, *args)
                }
                if (tr == null) {
                    Log.println(level, tag, msgFull)
                } else {
                    Log.println(level, tag, msgFull + "\n" + Log.getStackTraceString(tr))
                }
                logToFile(tag, msgFull, tr)
            }
        }

        private fun showLog(level: Int, tag: String): Boolean {
            return isLoggable(tag, level) || isLogEnabled
        }

        @SuppressLint("LogNotTimber")
        private fun isLoggable(tag: String, level: Int): Boolean {
            try {
                return Log.isLoggable(tag, level)
            } catch (e: Exception) {
                if (isLogEnabled) {
                    throw e
                } else {
                    Log.e(TAG, "please check the tag length?", e)
                }
            }
            return false
        }

        private fun logToFile(tag: String, msg: String?, tr: Throwable?) {
            if (isLogEnabled && fileLogger != null) {
                fileLogger!!.logToFile(tag, msg, tr)
            }
        }
    }
}
